1_benchmark_kvikIOzarr.py

```diff
@@ -0,0 +1,182 @@
+            ),
+            "unixtime1": torch.as_tensor(
+                data=sample.time.isel(time=1).astype(dtype="int64").data, device="cuda"
+            ),
+        }
+        for sample in samples
+    ]
+
+    return tensor_t0, tensor_t1, metadata
+
+
+# %%
+class WeatherBench2DataModule(L.LightningDataModule):
+    """
+    LightningDataModule to load WeatherBench2 data from Zarr.
+    """
+
+    def __init__(
+        self,
+        zarr_store: str = "2020-full_37-6h-0p25deg-chunk-1_zuv500.zarr",
+        # zarr_store: str = "gs://weatherbench2/datasets/era5/1959-2022-full_37-6h-0p25deg-chunk-1.zarr-v2",
+        engine: str = "kvikio",
+        batch_size: int = 32,
+    ):
+        """
+        Go from a Zarr datacube to 6-hourly time-slice chips!
+
+        Also does mini-batching.
+
+        Parameters
+        ----------
+        zarr_stores : str
+            A path or URL to a Zarr stores to read from. E.g. ``store1.zarr``.
+            See list of available WeatherBench2 Zarr stores at
+            https://weatherbench2.readthedocs.io/en/latest/data-guide.html
+
+        engine : str
+            The engine to use in `xr.open_dataset`. E.g. ``kvikio``, ``zarr``.
+            Default is ``kvikio``.
+
+        batch_size : int
+            Size of each mini-batch. Default is 32.
+
+        Returns
+        -------
+        datapipe : torchdata.datapipes.iter.IterDataPipe
+            A torch DataPipe that can be passed into a torch DataLoader.
+        """
+        super().__init__()
+        self.zarr_store: str = zarr_store
+        self.engine: str = engine
+        self.batch_size: int = batch_size
+
+        print(f"Loading data using {self.engine} engine")
+
+    def setup(self, stage: str | None = None) -> torchdata.datapipes.iter.IterDataPipe:
+        """
+        Data operations to perform on every GPU.
+        Split data into training and test sets, etc.
+
+        Returns
+        -------
+        datapipes : IterDataPipe
+            The torch DataPipe object to iterate over the training set.
+        """
+        # Step 0 - Iterate through all the Zarr stores
+        dp_source: torchdata.datapipes.iter.IterDataPipe = (
+            torchdata.datapipes.iter.IterableWrapper(iterable=[self.zarr_store])
+        )
+
+        # Step 1 - Create WeatherBench2 xarray.Dataset chips (full metadata)
+        dp_weather_chips: torchdata.datapipes.iter.IterDataPipe = (
+            # Step 1.0 - Open each Zarr store using xarray
+            dp_source.read_from_xpystac(
+                engine=self.engine, chunks=None, consolidated=False
+            )
+            # Step 1.1 - Select desired data variables at 500hPa
+            .map(fn=sel_datavars)
+            # Step 1.2 - Slice datacube along time-dimension into 12 hour chunks (2x 6-hourly)
+            .slice_with_xbatcher(
+                input_dims={"latitude": 721, "longitude": 1440, "time": 2},
+                preload_batch=False,
```

Comment:
Seems like `kvikio` engine (11s) is slightly faster than `zarr` (14s) when `preload_batch=False`, whereas `zarr` (10s) is slightly faster than `kvikio` (11s) when the default `preload_batch=True` is set. Maybe because the loading from dask.Array objects is not so optimized for kvikIO compared to Zarr yet? For benchmark purposes though, it's probably best to disable this `preload_batch` setting since it's somewhat like a cache, and we want to look at raw IO speed.